motion split with more flexibility

pi-r-p opened the pull request:
Slight change to introduce more flexibility...


## do nothing
```
$gts 
{ }
MOTIONSPLIT
```
![image](https://user-images.githubusercontent.com/37228740/68405346-3b446080-0180-11ea-9310-456d51e07c8b.png)

## time split, more than 10.95s between datapoints
```
$gts 
{ 
 'timesplit' 10.95 s 
}
MOTIONSPLIT
```
![image](https://user-images.githubusercontent.com/37228740/68405569-a2621500-0180-11ea-81bd-566fd9453f79.png)



## distance split, 100 m trigger
```
$gts 
{ 
 'timesplit' 10.95 s 
 'distance.split' 100
}
MOTIONSPLIT
```
![image](https://user-images.githubusercontent.com/37228740/68405544-95452600-0180-11ea-9c5f-303fee290218.png)

## combine, and add extra labels to output
```
$gts 
 //1573135239233178 1160575178 TIMECLIP // cut the final stop
{
  'timesplit' 10.95 s 
  'distance.split' 100
  'label.split.start' 'split'
  'label.split.type' 'reason'
}
MOTIONSPLIT
```
![image](https://user-images.githubusercontent.com/37228740/68405645-c4f42e00-0180-11ea-80be-b4c05d3d2912.png)

## stop detection split, simplest (time in an area), stop time label

```
$gts 
1573135239233178 1160575178 TIMECLIP // cut the final stop
{
  'stopped.min.time'  60 s
  'stopped.max.radius' 80
  'label.stopped.time' 'stop_time'
  'label.split.start' 'split'
  'label.split.type' 'reason'
}
MOTIONSPLIT
```
![image](https://user-images.githubusercontent.com/37228740/68405903-3207c380-0181-11ea-9d85-a7827e0681b9.png)

## stop detection split, with a maximum speed to consider stop
why ? Because when driving again, there is this extra point in the proximity zone, 
![image](https://user-images.githubusercontent.com/37228740/68405946-42b83980-0181-11ea-9b56-06150733a4f8.png)
```
<% 3.6 / %> 'km/h' STORE
$gts 
{
  'stopped.min.time'  60 s
  'stopped.max.speed' 10 @km/h
  'stopped.max.radius' 80
  'label.stopped.time' 'stop_time'
  'label.split.start' 'split'
  'label.split.type' 'reason'
}
MOTIONSPLIT
```
now it is OK:
![image](https://user-images.githubusercontent.com/37228740/68406188-a5a9d080-0181-11ea-83df-4c696da1a078.png)

## stop detection split, with a maximum mean speed in the proximity zone that prevent the split

It will split only if there is a full stop during the given time and radius, and if the mean speed in the proximity zone is less than 1km/h
```
$gts
{
  'stopped.min.time'  60 s
  'stopped.max.speed' 10 @km/h
  'stopped.max.radius' 80
  'stopped.max.mean.speed' 1 @km/h
  
  'label.stopped.time' 'stop_time'
  'label.split.start' 'split'
  'label.split.type' 'reason'
}
MOTIONSPLIT
```
![image](https://user-images.githubusercontent.com/37228740/68406570-557f3e00-0182-11ea-85b3-8a449b628d60.png)

## combining everything
```
$gts 
{
  'timesplit' 10.95 s 
  'distance.split' 100
  'stopped.min.time'  60 s
  'stopped.max.speed' 10 @km/h
  'stopped.max.radius' 80
  'stopped.max.mean.speed' 5 @km/h
  'label.stopped.time' 'stop_time'
  'label.split.start' 'split'
  'label.split.type' 'reason'
}
MOTIONSPLIT
```
![image](https://user-images.githubusercontent.com/37228740/68406738-98411600-0182-11ea-9f27-2b639c1192ca.png)


## purpose of stopped time in label
The label.stopped.time add an extra label with the time stopped. It is easy to split again later to isolate the stopped state. If you want, I can add this feature in the MOTIONSPLIT too.
```
$gts 
{
  'stopped.min.time'  60 s
  'stopped.max.speed' 10 @km/h
  'stopped.max.radius' 80
  'stopped.max.mean.speed' 5 @km/h
  'label.stopped.time' 'stop_time'
  'label.split.start' 'split'
  'label.split.type' 'reason'
}
MOTIONSPLIT

<%
  DROP
  'g' STORE
  $g LABELS 'stop_time' GET 'ts' STORE
  <% $ts ISNULL ! %>
  <%
    $g LASTTICK  $ts TOLONG - 'break' STORE
    $g [ [ MINLONG $break ]  [ $break MAXLONG ] ] CLIP
    DUP 1 GET { 'state' 'STOP' } RELABEL DROP 
    DUP 0 GET { 'state' 'RUN' } RELABEL DROP 
  %> 
  <%
    $g
  %>
  IFTE
%>
LFLATMAP
```
![image](https://user-images.githubusercontent.com/37228740/68407022-1ef5f300-0183-11ea-9594-7b70b0026c99.png)

## Split reason priority
Stopped > distance split > time split > end of sequence

pi-r-p committed:
motion split with more flexibility

hbs commented on warp10/src/main/java/io/warp10/script/functions/MOTIONSPLIT.java:
```diff
@@ -32,174 +33,303 @@
+    params = (Map) top;
+
+    long timeThreshold = Long.MAX_VALUE;
+    if (params.containsKey(PARAM_TIMESPLIT)) {
```
You can simplify by doing o = params.get(XXX) and then throwing an Exception if o is not of the expected type, since you don't support null as a valid value for any of the parameters

pi-r-p replied:
I don't want to throw an exception an a missing key, I just keep the default value set earlier.

hbs commented on warp10/src/main/java/io/warp10/script/functions/MOTIONSPLIT.java:
```diff
@@ -32,174 +33,303 @@
+      // If the previous tick was more than 'timeThreshold' ago, split now
+      //
+      if (timestamp - previousTimestamp > timeThreshold) {
+        if (null != splitTypeLabel) {
```
Simply set mustSplit to true and a splitReason to the reason for splitting, then set the label at the end of the loop.

pi-r-p replied:
done

hbs commented on warp10/src/main/java/io/warp10/script/functions/MOTIONSPLIT.java:
```diff
@@ -32,174 +33,303 @@
-      
-      if (timestamp - previousTimestamp > timeThreshold
-          || (GeoTimeSerie.NO_LOCATION != previousLocation && GeoTimeSerie.NO_LOCATION != location && GeoXPLib.orthodromicDistance(location, previousLocation) > distanceThreshold)) {
+      if (GeoTimeSerie.NO_LOCATION != previousLocation && GeoTimeSerie.NO_LOCATION != location && GeoXPLib.orthodromicDistance(location, previousLocation) > distanceThreshold) {
```
If the previous check lead to a split, should this one override the splitting reason?

pi-r-p replied:
Yes. In the current code, priority is:
Stopped>distance split>time split>end of sequence

Do you want to change this ?

pi-r-p replied:
I want to keep the priority of "stopped" detection over time split or distance split. That is why I do not do a if(time split) else if (distance split) else if (stop detection)

We can implement a DISTANCESPLIT function to have a fast one.

hbs commented on warp10/src/main/java/io/warp10/script/functions/MOTIONSPLIT.java:
```diff
@@ -32,174 +33,303 @@
-        if (previousTimestamp - refTimestamp > proximityZoneTime && proximityZoneTraveledDistance / ((double) (previousTimestamp - refTimestamp) / Constants.TIME_UNITS_PER_S) < proximityZoneSpeed) {
-          mustSplit = true;
+      long timeStopped = previousTimestamp - refTimestamp;
+      if (GeoTimeSerie.NO_LOCATION != refLocation && GeoTimeSerie.NO_LOCATION != location && null != split) {
```
Why checking for nullity of split here and not for the other tests?

pi-r-p replied:
Because I write the stopped time label later. It avoids an exception on a specific case:

```
NEWGTS
1 1 1 1 T ADDVALUE
{
  'stopped.min.time'  0 s
  'stopped.max.speed' 0 
  'stopped.max.radius' -1
  'label.stopped.time' 'stop_time'
  'label.split.start' 'split'
  'label.split.type' 'reason'
}
MOTIONSPLIT
```

hbs commented on warp10/src/main/java/io/warp10/script/functions/MOTIONSPLIT.java:
```diff
@@ -32,174 +33,303 @@
-        if (GeoTimeSerie.NO_LOCATION != previousLocation) {
-          proximityZoneTraveledDistance += GeoXPLib.orthodromicDistance(previousLocation, location);
+
+        // quit the radius, or speed greater than min speed.
```
min speed or max speed?

pi-r-p replied:
comment fixed.

hbs commented on warp10/src/main/java/io/warp10/script/functions/MOTIONSPLIT.java:
```diff
@@ -32,174 +33,303 @@
+          if (previousTimestamp != refTimestamp) {
+            zoneMeanSpeed = proximityZoneTraveledDistance / ((double) (previousTimestamp - refTimestamp) / Constants.TIME_UNITS_PER_S);
+          }
+          if (timeStopped > proximityZoneTime && zoneMeanSpeed < proximityInZoneMaxMeanSpeed) {
```
Actually this should be checked even if still in the proximity zone

pi-r-p replied:
High deceleration case: stops in less than radius : On the next iteration, the mean speed will be high too. That's why I added a separate max stopped speed, which is always checked.
What could be done next is to trig on the derivate (accerelation)

pi-r-p commented:
Do you want me to implement the "split run and stopped phase" function ? The one I did in WarpScript with an LFLATMAP upper.

pi-r-p committed:
motion split review

pi-r-p committed:
introduce stopped/moving split, replace split start by split number, fix empty gts or gts with a single point

pi-r-p committed:
fix the one point moving when stopped at the beginning

pi-r-p committed:
fix one point overlap on moving/stopped split

pi-r-p commented:
- Introduction of 'stopped.split' boolean. If true, all the split created by a stop detection will be splitted again in a moving split and a stopped split.
```
$gts 
//1573135340837236 1262179236 TIMECLIP // cut the final stop
//1573135585347000 1364277854 TIMECLIP
{
  'stopped.min.time'  60 s
  'stopped.max.speed' 10 @km/h
  'stopped.max.radius' 80
  'stopped.split' true
  'label.split.number' 'split'
  'label.split.type' 'reason'
}
MOTIONSPLIT
```
![image](https://user-images.githubusercontent.com/37228740/68484737-28906100-023e-11ea-99d4-ead75958ee0c.png)

- To be consistent with current TIMESPLIT, I replaced the start timestamp label with a split number (starting from 1, like TIMESPLIT)

hbs commented on warp10/src/main/java/io/warp10/script/functions/MOTIONSPLIT.java:
```diff
@@ -32,174 +33,371 @@
+      //
+      // If the previous tick was more than 'timeThreshold' ago, split now
+      //
+      if (timestamp - previousTimestamp > timeThreshold) {
```
Can chain the tests in if ... else if ... else if constructs in the order you want (probably the reverse order of the current list of ifs) so we do not execute useless code

pi-r-p replied:
The most complex one (proximity zone detection) must be done on each iteration. I added a test on proximity radius and time, to save time in case of a distance or time only detection.
Then moved time and distance tests in a if ... else if (with comments to explain priorities)

Unit tests needed before merge.

pi-r-p committed:
optimize the tests for speed if no proximity zone detection

pi-r-p committed:
fixed missing 'end' splitTypeLabel

hbs commented on warp10/src/main/java/io/warp10/script/functions/MOTIONSPLIT.java:
```diff
@@ -32,174 +33,376 @@
+        //
+        // Split again the current split if proximityZoneSplit is set
+        //
+        if (proximityZoneSplit && null != split && splitReason == SPLIT_TYPE_STOPPED) {
```
Use SPLIT_TYP_STOPPED.equals(splitReason), in your case this is equivalent, but comparing strings using == is errorprone most of the time

pi-r-p committed:
fix string equality

pi-r-p committed:
fixed the reason label when proximitySplit set

pi-r-p commented:
Documentation and unit tests ready. Could be merged.
